packages/vendure-plugin-shipmate/src/api/shipmate.service.ts

```diff
@@ -0,0 +1,271 @@
+    ] = shipmateConfig.apiKey;
+  }
+
+  async getShipmentToken(
+    shipmateUsername: string,
+    shipmatePassword: string
+  ): Promise<string> {
+    const response = await firstValueFrom(
+      this.httpService.post<GetTokenRespose>(
+        `${this.config.shipmateApiUrl}/tokens`,
+        {
+          username: shipmateUsername,
+          password: shipmatePassword,
+        }
+      )
+    );
+    if (response.data.data?.token) {
+      Logger.info('Successfully authenticated with Shipmate API', loggerCtx);
+      return response.data.data.token;
+    } else {
+      Logger.error(response.data.message, loggerCtx);
+      throw new UserInputError(response.data.message);
+    }
+  }
+
+  async createShipment(ctx: RequestContext, order: Order) {
+    const shipmateReference = generatePublicId();
+    const payload = parseOrder(order, shipmateReference);
+    try {
+      const result = await firstValueFrom(
+        this.httpService.post<CreateShipmentResponse>(
+          `${this.config.shipmateApiUrl}/shipments`,
+          payload
+        )
+      );
+      Logger.info(result.data.message, loggerCtx);
+      await this.orderService.updateCustomFields(ctx, order.id, {
+        ...order.customFields,
+        shipmateReference,
+      });
+    } catch (error: any) {
+      const message = error.response.data.message;
+      Logger.error(message, loggerCtx);
+    }
+  }
+
+  async updateOrderState(payload: EventPayload, res: Response): Promise<void> {
+    const ctx = await this.createCtx(payload.auth_token);
+    if (!ctx) {
+      const message = `No registered ShipmentConfigEntity with this auth_token`;
+      Logger.error(message, loggerCtx);
+      res.status(HttpStatus.BAD_REQUEST).json({ message });
+      return;
+    }
+    await this.connection.startTransaction(ctx);
+    const shipmentOrder = await this.orderService.findOneByCode(
+      ctx,
+      payload.order_reference
+    );
+    if (!shipmentOrder) {
+      const message = `No Order with code ${payload.order_reference}`;
+      Logger.error(message, loggerCtx);
+      res.status(HttpStatus.BAD_REQUEST).json({ message });
+      return;
+    }
+    Logger.info(
+      `${payload.event} event received for Order with code ${payload.order_reference}`,
+      loggerCtx
+    );
+    if (payload.event === 'TRACKING_COLLECTED') {
+      await this.updateFulFillment(ctx, shipmentOrder, payload, 'Shipped');
+      res
+        .status(HttpStatus.CREATED)
+        .json({ message: `Order marked as  Shipped successfully` });
+    } else if (payload.event === 'TRACKING_DELIVERED') {
+      await this.updateFulFillment(ctx, shipmentOrder, payload, 'Delivered');
+      res
+        .status(HttpStatus.CREATED)
+        .json({ message: `Order marked as Delivered successfully` });
+    } else {
+      Logger.info(
+        `No configured handler for event "${payload.event}"`,
+        loggerCtx
+      );
+    }
+    await this.connection.commitOpenTransaction(ctx);
+    return;
```

Comment:
`res` is pretty `controller` specific, but a service should only contain business logic, and not really be aware of HTTP or Graphql. A better approach is to throw an error here (UserInputError, ForbiddenError etc), and let the controller decide how this should be returned as HTTP error.

Considering this is a webhook, it's also fine to just differentiate between Error or returning. The different HTTP return statusses aren't used by the shipmate webhook i think? They only care about OK or not, and retry if not OK.

Reply:
Why are we manually starting and commiting transactions?